crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -566,6 +616,155 @@ impl Object {
 
         Ok(conn)
     }
+
+    async fn query_live(db: &Db, address: SuiAddress) -> Result<Option<Self>, Error> {
+        let vec_address = address.into_vec();
+        use objects::dsl as objects;
+
+        let stored_obj: Option<StoredObject> = db
+            .execute(move |conn| {
+                conn.result(move || {
+                    objects::objects
+                        .filter(objects::object_id.eq(vec_address.clone()))
+                        .limit(1)
+                        .into_boxed()
+                })
+                .optional()
+            })
+            .await
+            .map_err(|e| Error::Internal(format!("Failed to fetch object: {e}")))?;
+
+        stored_obj.map(Self::try_from).transpose()
+    }
+
+    async fn query_historical(
+        db: &Db,
+        address: SuiAddress,
+        version: Option<u64>,
+        checkpoint_sequence_number: Option<u64>,
+    ) -> Result<Option<Self>, Error> {
+        let vec_address = address.into_vec();
+        let version = version.map(|n| n as i64);
+        let checkpoint_sequence_number = checkpoint_sequence_number.map(|n| n as i64);
+
+        use checkpoints::dsl as checkpoints;
+        use objects_history::dsl as history;
+        use objects_snapshot::dsl as snapshot;
+
+        let results: Option<Vec<StoredHistoryObject>> = db
+            .execute(move |conn| {
+                conn.results(move || {
+                    // If an object was created or mutated in a checkpoint outside the current
+                    // available range, and never touched again, it will not show up in the
+                    // objects_history table. Thus, we always need to check the objects_snapshot
+                    // table as well.
+                    let mut snapshot_query = snapshot::objects_snapshot
+                        .filter(snapshot::object_id.eq(vec_address.clone()))
+                        .into_boxed();
+
+                    let mut historical_query = history::objects_history
+                        .filter(history::object_id.eq(vec_address.clone()))
+                        .order_by(history::object_version.desc())
+                        .limit(1)
+                        .into_boxed();
+
+                    if let Some(version) = version {
+                        snapshot_query =
+                            snapshot_query.filter(snapshot::object_version.eq(version));
+                        historical_query =
+                            historical_query.filter(history::object_version.eq(version));
+                    }
+
+                    let left = snapshot::objects_snapshot
+                        .select(snapshot::checkpoint_sequence_number)
+                        .order(snapshot::checkpoint_sequence_number.desc())
+                        .limit(1);
+
+                    historical_query = historical_query.filter(
+                        history::checkpoint_sequence_number.ge(coalesce(left.single_value(), 0)),
```

Comment:
Do you need `coalesce` for this? I would really like to avoid adding it if we can, because it's postgres specific.  Can we try adding a `left.single_value().is_not_null()` filter instead?